feat!: split private event emission from delivery

nventuro opened the pull request:
The meat of this change is the following:

```diff
- self.emit(event, recipient, delivery_method)
+ self.emit(event).delivery(recipient, delivery_method)
```

This clears lots of things internally, exposing concepts that were hidden inside functions. For end users, it most notably provides a way for the same event to be delivered to multiple recipients, which was impossible before. It also nicely mirrors how notes are handled, which should help these concepts land better.

There's multiple things that are wrong about events right now, both in terms of the code structure and in terms of behavior (e.g. we don't commit to the event type atm), but I wanted to first get this changeset out so that we can include it in the next release and get it in the hands of users.

benesjan commented on noir-projects/aztec-nr/aztec/src/event/event_interface.nr:
```diff
@@ -1,5 +1,25 @@
     fn get_event_type_id() -> EventSelector;
 }
+
+/// A private event's commitment is used to validate that an event was indeed emitted.
```
```suggestion
/// A private event's commitment is stored on-chain and is used to validate that an event contained in a
/// message was indeed emitted by the expected contract.
```

I think it might not be super clear when this validation is supposed to take place.

benesjan commented on noir-projects/aztec-nr/aztec/src/contract_self.nr:
```diff
@@ -175,34 +179,44 @@ impl<Storage, CallSelf, EnqueueSelf, CallSelfStatic, EnqueueSelfStatic, CallInte
-    /// - `recipient`: The address that should be able to decrypt and read this event
-    /// - `delivery_mode`: The delivery mode for the event (e.g., `MessageDelivery.CONSTRAINED_ONCHAIN`)
+    /// Unlike public events, private events do not reveal their contents publicly. They instead create an
+    /// [EventMessage] containing the private event information, which **MUST** be delivered to a recipeint via
```
```suggestion
    /// [EventMessage] containing the private event information, which **MUST** be delivered to a recipient via
```

mverzilli replied:
```suggestion
    /// [EventMessage] containing the private event information, which **MUST** be delivered to a recipient via
```

benesjan commented on noir-projects/noir-contracts/contracts/app/token_contract/src/main.nr:
```diff
@@ -243,7 +243,10 @@ pub contract Token {
         // function is only designed to be used in situations where the event is not strictly necessary (e.g. payment to
         // another person where the payment is considered to be successful when the other party successfully decrypts a
         // note).
```
Unrelated to the PR but this comment is out of place as it talks about notes and is above event message. Would move it up in the function and talk about general messages and not note message.

benesjan reviewed (approved):
Very nice

nventuro committed:
feat!: split private event emission from delivery

The meat of this change is the following:

```diff
- self.emit(event, recipient, delivery_method)
+ self.emit(event).delivery(recipient, delivery_method)
```

This clears lots of things internally, exposing concepts that were hidden inside functions. For end users, it most notably provides a way for the same event to be delivered to multiple recipients, which was impossible before. It also nicely mirrors how notes are handled, which should help these concepts land better.

There's multiple things that are wrong about events right now, both in terms of the code structure and in terms of behavior (e.g. we don't commit to the event type atm), but I wanted to first get this changeset out so that we can include it in the next release and get it in the hands of users.
